Add random partitioning on SageMaker. Update run and compose scripts along with instructions.

thvasilo opened the pull request:
*Description of changes:*
Add random partitioning on SageMaker, and support for partitioning using Docker compose.

* The SM Dockerfile is changed to clone DGL as we use `tools/data_dispatch.py` and install all the necessary dependencies.
* We change the way we generate Docker compose files, using a sub-parser with an action for each of `partitioning`, `training`, `inference`. This helps create better separation and logical grouping of arguments.
* We add a general/extensible way to run distributed partition algorithms on SageMaker, with initial support for random partitioning.

Most of the code changes add the new code for partitioning. I would suggest focusing on changed code rather than the new files to see the new way to run `generate_sagemaker_docker_compose.py`, that is e.g.:

```bash
GSF_ARGS="--num-layers 1 --hidden-size 128 --backend gloo --batch-size 128"
python local/generate_sagemaker_docker_compose.py --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    training \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

Most of the changes in the README are about explaining the changes to Docker compose.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

thvasilo commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+
+        data_dispatch_step(local_partition_path)
+
+        # Indicate we can stop sending keepalive messages
+        task_end.set()
+        # Ensure the keepalive thread has finished before closing sockets
+        thread.join()
+        # Close connections with workers
+        utils.terminate_workers(client_list, world_size, task_end)
```
Rest of executions scripts need to do the same.

classicsong commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+@dataclass
+class PartitionJobConfig():
+    """
+    Configuration object for a SageMaker partitioning job.
```
Can you have a description of each field here?

classicsong commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+        edge_file_list = edge_data["data"]
+        read_list = np.array_split(np.arange(len(edge_file_list)), world_size)
+        for i, efile in enumerate(edge_file_list):
+            # TODO: Only download round-robin if ParMETIS will run, skip otherwise
```
What about random partition?

classicsong commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+        for i, efile in enumerate(edge_file_list):
+            # TODO: Only download round-robin if ParMETIS will run, skip otherwise
+            # Download files both in round robin and sequential assignment
+            if i % world_size == local_rank or i in read_list[local_rank]:
```
Why both in round robin and sequential?
It is weird here.

thvasilo replied:
ParMETIS assumes round-robin assignment, while data_dispatch assume sequential. So we need to have both available locally. We could remove the round-robin for now since we only have random though.

thvasilo replied:
I'll have a follow-up PR that only downloads in the way that data_dispatch expects, would that cover it?

classicsong commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+
+    logging.info("Downloading graph data from %s into %s",
+        graph_data_s3, tmp_data_path)
+    graph_data_path = download_graph(
```
if skip_partitioning, do we need to download the graph data?

thvasilo replied:
Yes, the local data are used by data_dispatch to create the DistDGL files right?

classicsong commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,371 @@
+        download_data_from_s3(s3_partition_path, local_partition_path, sagemaker_session)
+    else:
+        # Perform the partitioning. Leader uploads partition assignments to S3.
+        local_partition_path, s3_partition_path = partitioner.create_partitions(
```
So now only leader do the random partition?

thvasilo replied:
For random partitioning yes we don't need to involve the workers as the partitioning takes seconds even for very large graphs. If we run into any performance issues we can revisit.

To be clear: `sagemaker_partition.py` is agnostic of the partitioning algorithm implementation, as long as it follows the contracts of the `PartitionAlgorithm` base class. 

One of the contracts is that at the end of the partition assignment step, the partition assignment files exist on the leader, which proceeds to upload them to the designated location.

classicsong commented on python/graphstorm/sagemaker/sagemaker_train.py:
```diff
@@ -150,8 +150,8 @@ def run_train(args, unknownargs):
 
-    print(f"Know args {args}")
-    print(f"Unknow args {unknownargs}")
+    print(f"Known args {args}")
```
Put these into a typo PR.

classicsong commented on python/graphstorm/sagemaker/utils.py:
```diff
@@ -206,16 +233,40 @@ def download_graph(graph_data_s3, graph_name, part_id, local_path, sagemaker_ses
+    s3_input_key = graph_data_s3.split("/", maxsplit=3)[3]
+
+    s3_client = boto3.client('s3')
+    for graph_config  in [f"{graph_name}.json", "metadata.json"]:
```
Why f"{graph_name}.json" and "metadata.json" here?

thvasilo replied:
data_dispatch.py creates a file named `metadata.json`. I guess the single-instance partitioning instead creates `{graph_name}.json` so we should support both.

classicsong commented on sagemaker/launch/launch_infer.py:
```diff
@@ -109,48 +108,58 @@ def run_job(input_args, image, unknowargs):
              "Do not store it with partitioned graph")
     parser.add_argument("--model-artifact-s3", type=str,
-        help="S3 bucket to load the saved model artifacts")
+        required=True,
```
Let's put it into another PR.

classicsong commented on sagemaker/launch/launch_train.py:
```diff
@@ -65,7 +64,7 @@ def run_job(input_args, image, unknowargs):
     container_image_uri = image
 
-    prefix = "script-mode-container"
+    job_prefix = input_args.task_name if input_args.task_name else f"train-{input_args.graph_name}"
```
Put it into another PR.

classicsong commented on sagemaker/run/infer_entry.py:
```diff
@@ -21,21 +21,29 @@
 
-def parse_train_args():
-    """  Add arguments for model training
+def get_inference_parser():
```
Put it into another PR.

thvasilo commented on docker/sagemaker/Dockerfile.sm:
```diff
@@ -13,6 +13,11 @@ RUN pip3 install dgl==1.0.2 -f https://data.dgl.ai/wheels/cu117/repo.html
 
+# Install MPI etc needed by DistDGL
+RUN apt-get update; apt-get install -y --no-install-recommends libopenmpi-dev \
+    build-essential make g++ software-properties-common; add-apt-repository ppa:ubuntu-toolchain-r/test; \
```
Good point, I took most of the deps from an image where I was building DGL. Let me try to minimize these.

thvasilo commented on python/graphstorm/sagemaker/partition_algorithm.py:
```diff
@@ -0,0 +1,232 @@
+        partition_dir = os.path.join(self.local_output_path, "partition")
+        os.makedirs(partition_dir, exist_ok=True)
+
+        # Random partitioning is done on the leader node only
```
The base class is written assuming distributed execution (e.g. ParMETIS). Here we only utilize the leader to do the partitioning assignments, but the caller and synchronization between stages still happens using the leader/worker scheme.

thvasilo commented on python/graphstorm/sagemaker/partition_algorithm.py:
```diff
@@ -0,0 +1,232 @@
+                [pa.array(partition_assignment, type=pa.int64())],
+                names=["partition_id"])
+            options = pa_csv.WriteOptions(include_header=False, delimiter=' ')
+            pa_csv.write_csv(arrow_partitions, ntype_output, write_options=options)
```
It's the expected input for DistDGL's `dispatch_data.py`

thvasilo commented on python/graphstorm/sagemaker/partition_algorithm.py:
```diff
@@ -0,0 +1,232 @@
+            raise RuntimeError("broadcast_partition_done should only be called by the leader")
+        msg = b"PartitionDone" if success else b"PartitionFail"
+        for rank in range(1, world_size):
+            client_list[rank].sendall(msg)
```
`sendall` is safer here right? It ensures the whole buffer is sent before returning.

thvasilo commented on python/graphstorm/sagemaker/s3_utils.py:
```diff
@@ -0,0 +1,55 @@
+    sagemaker_session: sagemaker.session.Session
+        sagemaker_session to run download
+    """
+    logging.debug("download %s into %s", input_s3, local_data_path)
```
I think we can leave it as the responsibility of the caller. If you think we need to take care of this I can add a simple check.

thvasilo commented:
Hi @classicsong I've addressed most of the comments here and resolved the conversations that I think were answered. If you think we need some more changes feel free to re-open one of the resolved comments and I'll address it.

classicsong commented on sagemaker/local/generate_sagemaker_docker_compose.py:
```diff
@@ -18,111 +18,224 @@
 """
```
Shall we have separate compose.py for partition and training/inference?

thvasilo replied:
My proposed design and implementation here uses one compose file, with different sub-parsers and commands. For example to generate a `training` compose you'd use:


```bash
python local/generate_sagemaker_docker_compose.py --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    training \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

See the README for details.

classicsong replied:
Can we do the following?
```
python local/generate_sagemaker_docker_compose.py  training \
    --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

thvasilo replied:
Done in latest commit.

thvasilo commented on python/graphstorm/sagemaker/sagemaker_partition.py:
```diff
@@ -0,0 +1,388 @@
+                # TODO: Use dgl.tools.distpartitioning.utils.generate_read_list
+                # once we move the code over from DGL
+                read_list = np.array_split(np.arange(num_files), world_size)
+                for i in read_list[local_rank].tolist():
```
This arrangement is sequential, e.g.

```python
In [188]: num_files = 8

In [189]: world_size = 4

In [190]: np.array_split(np.arange(num_files), world_size)
Out[190]: [array([0, 1]), array([2, 3]), array([4, 5]), array([6, 7])]
```

classicsong replied:
I c.

thvasilo commented on docker/sagemaker/Dockerfile.sm:
```diff
@@ -24,6 +29,13 @@ ENV PYTHONPATH="/opt/ml/code/graphstorm/python/:${PYTHONPATH}"
 RUN cp /opt/ml/code/graphstorm/sagemaker/run/* /opt/ml/code/
 
+# Download DGL source code
+RUN cd /root; git clone https://github.com/dmlc/dgl.git
```
Yeah, let me do that and add it.
